plugins/woocommerce/src/Internal/Utilities/BatchProcessingController.php

```diff
@@ -0,0 +1,287 @@
+	 * @param \Exception|null $last_error Exception object in processing the batch, if there was one.
+	 */
+	protected function update_progress_status( BatchProcessor $batch_processor, array $batch, float $time_taken, \Exception $last_error = null ) {
+		$current_status                      = $this->get_process_details( $batch_processor );
+		$current_status['total_time_spent'] += $time_taken;
+		$current_status['last_processed']    = null !== $last_error ? end( $batch ) : $current_status['last_processed'];
+		$current_status['last_error']        = null !== $last_error ? $last_error->getMessage() : null;
+		update_option( $this->get_process_option_name( $batch_processor ), $current_status, false );
+	}
+
+	/**
+	 * Scheduler next update batch.
+	 *
+	 * @param string $process_name Fully qualified class name of the processor.
+	 *
+	 * @return int Action ID.
+	 */
+	private function schedule_next_batch( string $process_name ) : int {
+		return as_schedule_single_action( time(), self::SINGLE_BATCH_PROCESS_ACTION, array( $process_name ) );
+	}
+
+	/**
+	 * Whether the given updater is already scheduled.
+	 * Differs from `as_has_scheduled_action` in that this excludes actions in progress.
+	 *
+	 * @param string $process_name Fully qualified class name of the batch processor.
+	 *
+	 * @return bool Whether it's already scheduled.
+	 */
+	public function already_scheduled( string $process_name ) {
+		return as_has_scheduled_action( self::SINGLE_BATCH_PROCESS_ACTION, array( $process_name ) );
+	}
+
+	/**
+	 * Get instance of updater for given class name,
+	 *
+	 * @param string $processor_class_name Class name of batch processor.
+	 *
+	 * @return BatchProcessor Instance of batch processor.
+	 * @throws \Exception If processor class doesn't exist.
+	 */
+	private function get_processor_instance( string $processor_class_name ) : BatchProcessor {
+		$processor = wc_get_container()->get( $processor_class_name );
+
+		/**
+		 * Filters the instance of processor for current class name.
+		 *
+		 * @since 6.7.0.
+		 */
+		$processor = apply_filters( 'woocommerce_get_batch_processor', $processor, $processor_class_name );
+		if ( ! isset( $processor ) && class_exists( $processor_class_name ) ) {
+			// This is a fallback for when the batch processor is not registered in the container.
+			$processor = new $processor_class_name();
+		}
+		if ( ! is_a( $processor, BatchProcessor::class ) ) {
+			throw new \Exception( 'Unable to initialize batch processor instance.' );
+		}
+		return $processor;
+	}
+
+	/**
+	 * Helper method to get list of all pending processes.
+	 *
+	 * @return array List of pending processes.
+	 */
+	public function get_pending() : array {
+		return get_option( self::PENDING_PROCESSES_OPTION_NAME, array() );
+	}
+
+	/**
+	 * Mark a pending process complete.
+	 *
+	 * @param string $processor Name of processor class which will be marked compete.
+	 */
+	protected function mark_pending_process_complete( string $processor ) {
```

Comment:
Again, could this perhaps use a hook instead of being a protected method?

Reply:
I have marked this method as private for now. As you said in an earlier conversation, cleanup action can be done after checking if there is no batch remaining.